### PR TITLE
feat: Named Graphs API

### LINKED ISSUE #12: Support for Named Graphs
As a jsonld4s library user, I'd like to be able to create Named Graphs as described [here](https://json-ld.org/spec/latest/json-ld/#named-graphs).

**Acceptance criteria:**
- [ ] the API should allow the creation of the following constructs:
  ```json
  {
    "@id": "_:graph",
    "@graph": [
      {
        "@id": "http://manu.sporny.org/about#manu",
        "@type": "http://schema.org/Person",
        "name": "Manu Sporny",
        "knows": "http://greggkellogg.net/foaf#me"
      }, {
        "@id": "http://greggkellogg.net/foaf#me",
        "@type": "http://schema.org/Person",
        "name": "Gregg Kellogg",
        "knows": "http://manu.sporny.org/about#manu"
      }
    ]
  }
  ```

**Notes:**
* the `@context` is descoped;
* the graph `@id` is mandatory; in the future, once `@context` is added, it should be possible to create a grouping without the `@id` so there's a possibility of a list of objects in the default graph;

### PR DESCRIPTION
**Pull request by jachro**
closes #12

### CODE REVIEW

File: src/main/scala/io/renku/jsonld/JsonLD.scala [hunk 1]
```diff
@@ -198,7 +209,32 @@ object JsonLD {
...
+    override type T = JsonLD
+
+    override lazy val merge: Either[MalformedJsonLD, T] = {
+
+      def isFlatten(jsons: Seq[JsonLD]): Boolean = jsons.exists {
+        case _: JsonLDEdge => true
+        case _ => false
+      }
+
+      lazy val collectEntityLikeEntities: Seq[JsonLD] => (List[JsonLD], List[JsonLDEntityLike]) =
```
Comment (eikek) [inline src/main/scala/io/renku/jsonld/JsonLD.scala:223]: It's too bad that there is no "typed span" on a `List`. I think cats has some utility for it 
```scala
_.partitionEither {
      case a: JsonEntityLike => Right(a)
      case a => Left(a)
    }
```
Ofc, it's the same thing at the end!

File: src/main/scala/io/renku/jsonld/NamedGraph.scala [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+  def fromJsonLDsUnsafe(id: EntityId, jsonLD: JsonLD, otherJsonLDs: JsonLD*): NamedGraph =
+    from(id, jsonLD, otherJsonLDs: _*).fold(throw _, identity)
+
+  @tailrec
+  private def flatten(toFlatten: List[JsonLD],
+                      flattened: List[JsonLDEntityLike]
+  ): Either[IllegalArgumentException, List[JsonLDEntityLike]] = toFlatten match {
+    case Nil                              => flattened.asRight
+    case (json: JsonLDEntityLike) :: more => flatten(more, flattened ::: json :: Nil)
```
Comment (eikek) [inline src/main/scala/io/renku/jsonld/NamedGraph.scala:69]: Maybe a mini-tuning could be to use a `::` to prepend the `json` and reverse the list at the end? But depends on the average size whether it is useful…

**Review by eikek — APPROVED** [review]
👍🏼

File: src/main/scala/io/renku/jsonld/JsonLD.scala [hunk 3]
```diff
@@ -97,13 +99,15 @@ object JsonLD {
   def edge(source: EntityId, property: Property, target: EntityId): JsonLDEdge =
     JsonLDEdge(source, property, target)
 
-  private[jsonld] final case class JsonLDEntity(id:         EntityId,
-                                                types:      EntityTypes,
-                                                properties: Map[Property, JsonLD],
-                                                reverse:    Reverse
-  ) extends JsonLD
+  sealed trait JsonLDEntityLike extends JsonLD
```
[the remark this answers is not in the document]
Comment (jachro) [reply src/main/scala/io/renku/jsonld/JsonLD.scala:102]: Yes, from one side, it's a public trait so a comment might be nice. However, from the other side I'd like to hide it as, let's say in the Scala 3 world, I'd use a sum type on all the places where I needed to use it as an argument to a `def`.

File: src/main/scala/io/renku/jsonld/JsonLD.scala [hunk 4]
```diff
@@ -21,13 +21,15 @@ package io.renku.jsonld
...
 import io.renku.jsonld.flatten.{JsonLDArrayFlatten, JsonLDEntityFlatten, JsonLDFlatten}
-import io.renku.jsonld.merge.{JsonLDArrayMerge, JsonLDMerge}
+import io.renku.jsonld.merge.{EntitiesMerger, JsonLDMerge}
 
 import java.io.Serializable
 import java.time.{Instant, LocalDate}
 
 abstract class JsonLD extends JsonLDMerge with JsonLDFlatten with Product with Serializable {
 
+  type T <: JsonLD
```
[the remark this answers is not in the document]
Comment (jachro) [reply src/main/scala/io/renku/jsonld/JsonLD.scala:31]: Hmm, the problem is that, as far as I remember :), the abstract type is used by both: `JsonLDMerge ` and `JsonLDFlatten`.

[the remark this answers is not in the document]
Comment (eikek) [reply src/main/scala/io/renku/jsonld/JsonLD.scala:31]: Ah I didn't see this!

**Review by eikek — APPROVED** [review]

**Pull request merged**